### PR TITLE
[EC-423] Fix unit tests

### PR DESCRIPTION
**Pull request by eliykat**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Fix a few unrelated issues we were having when running tests.

## Code changes

* move partial jest config from `package.json` to `jest.config.ts`. You can use one location or the other, but not both.
* fix broken import in `libs/common/spec/models/domain/encArrayBuffer.spec.ts`
* fix implementation of `bitButton`. Per [this test](https://github.com/bitwarden/clients/blob/master/libs/components/src/button/button.directive.spec.ts#L42-L45), setting `buttonType = null` should default to the secondary type, but this test is currently failing. Fix how we handle the default buttonType.
* fix Firefox import data - `baseImporter` was logging parsing errors to the console every time tests were run, presumably from malformed test data. This probably meant that the test was passing for the wrong reason. Replacing this line with fresh (redacted) export data fixes it.

## Before you submit

```
- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

**Review by djsmith85 — DISMISSED** [review]
@eliykat Thank you for fixing these. I also noticed some broken tests earlier today. 

I can approve all change besides the button.directive. I was not part of that initiative. Will leave that up to EC and SSG

File: jest.config.js [hunk 1]
```diff
@@ -19,4 +19,5 @@ module.exports = {
     "<rootDir>/libs/electron/jest.config.js",
     "<rootDir>/libs/node/jest.config.js",
   ],
+  testEnvironment: "node",
```
Comment (MGibson1) [inline jest.config.js:22]: I don't think this appropriate here, I'd rather set this in the individual project files, if necessary.

It also defaults to node, so not necessary

**Review by MGibson1 — APPROVED** [review]
I don't like the `testEnvironment = node`, but I can't see that it hurts anything

**Comment by eliykat** [comment]
I've made the following additional changes. This is all to enable #3216 so that we can have unit tests run in the pipeline again.

* create base jest config which all projects can extend, to minimise duplication
* add `.spec` files to web `tsconfig` so that we have proper path mapping (enabling the `@bitwarden/common` imports)
* fix a few more tests which have been broken on `master`
* add web to the root `jest.config` so that it's included when all tests are run
* undo the `testEnvironment` change @MGibson1 commented on (I discussed with Vince)
* add `maxWorkers: 3` setting, see comment for explanation

I've forked this and merged #3216 into it to confirm that it successfully runs all tests: https://github.com/eliykat/bitwarden-clients/runs/7759609192?check_suite_focus=true

File: apps/cli/jest.config.js [hunk 2]
```diff
@@ -2,13 +2,10 @@ const { pathsToModuleNameMapper } = require("ts-jest");
 
 const { compilerOptions } = require("./tsconfig");
 
+const sharedConfig = require("../../libs/shared/jest.config.base");
+
 module.exports = {
   preset: "ts-jest",
-  testMatch: ["**/+(*.)+(spec).+(ts)"],
-  setupFilesAfterEnv: ["<rootDir>/spec/test.setup.ts"],
```
Comment (MGibson1) [inline apps/cli/jest.config.js:8]: I realize this is empty, but I would still like to keep the file around for if/when we need setup.

File: apps/browser/tsconfig.json [hunk 3]
```diff
@@ -17,5 +17,6 @@
   "angularCompilerOptions": {
     "preserveWhitespaces": true
   },
-  "include": ["src"]
+  "include": ["src"],
+  "files": ["./test.setup.ts"]
```
Comment (MGibson1) [inline apps/browser/tsconfig.json:21]: we have `tsconfig.spec.json` to configure test includes in, I don't think we want to include any specs or test setups in the base `tsconfig` files.

Happy to be wrong about that, but I don't want tests in our prod outputs and I'm not sure they won't be if they are included here.

File: apps/web/tsconfig.json [hunk 4]
```diff
@@ -15,6 +15,11 @@
...
   },
-  "files": ["src/polyfills.ts", "src/main.ts", "../../bitwarden_license/bit-web/src/main.ts"],
-  "include": ["src/connectors/*.ts", "src/**/*.stories.ts"]
+  "files": [
+    "src/polyfills.ts",
+    "src/main.ts",
+    "../../bitwarden_license/bit-web/src/main.ts",
+    "./test.setup.ts"
+  ],
+  "include": ["src/connectors/*.ts", "src/**/*.stories.ts", "src/**/*.spec.ts"]
```
Comment (MGibson1) [inline apps/web/tsconfig.json:24]: we have `tsconfig.spec.json` to configure test includes in, I don't think we want to include any specs or test setups in the base `tsconfig` files.

Happy to be wrong about that, but I don't want tests in our prod outputs and I'm not sure they won't be if they are included here.

File: libs/angular/tsconfig.json [hunk 5]
```diff
@@ -1,5 +1,6 @@
 {
   "extends": "../shared/tsconfig.libs",
   "include": ["src", "spec"],
-  "exclude": ["node_modules", "dist"]
+  "exclude": ["node_modules", "dist"],
+  "files": ["./test.setup.ts"]
```
Comment (MGibson1) [inline libs/angular/tsconfig.json:5]: we have `tsconfig.spec.json` to configure test includes in, I don't think we want to include any specs or test setups in the base `tsconfig` files.

Happy to be wrong about that, but I don't want tests in our prod outputs and I'm not sure they won't be if they are included here.

File: libs/common/jest.config.js [hunk 6]
```diff
@@ -6,11 +6,7 @@ module.exports = {
   displayName: "libs/common tests",
   preset: "ts-jest",
   testEnvironment: "jsdom",
-  testMatch: ["**/+(*.)+(spec).+(ts)"],
-  setupFilesAfterEnv: ["<rootDir>/spec/test.setup.ts"],
-  collectCoverage: true,
-  coverageReporters: ["html", "lcov"],
-  coverageDirectory: "coverage",
+  setupFilesAfterEnv: ["<rootDir>/test.setup.ts"],
```
Comment (MGibson1) [inline libs/common/jest.config.js:9]: This seems to be missing your shared setup

File: libs/common/tsconfig.json [hunk 7]
```diff
@@ -1,5 +1,6 @@
 {
   "extends": "../shared/tsconfig.libs",
   "include": ["src", "spec"],
-  "exclude": ["node_modules", "dist"]
+  "exclude": ["node_modules", "dist"],
+  "files": ["./test.setup.ts"]
```
Comment (MGibson1) [inline libs/common/tsconfig.json:5]: we have `tsconfig.spec.json` to configure test includes in, I don't think we want to include any specs or test setups in the base `tsconfig` files.

Happy to be wrong about that, but I don't want tests in our prod outputs and I'm not sure they won't be if they are included here.

File: libs/components/tsconfig.json [hunk 8]
```diff
@@ -28,5 +28,6 @@
     "strictInjectionParameters": true,
     "strictInputAccessModifiers": true,
     "strictTemplates": true
-  }
+  },
+  "files": ["./test.setup.ts"]
```
Comment (MGibson1) [inline libs/components/tsconfig.json:32]: we have `tsconfig.spec.json` to configure test includes in, I don't think we want to include any specs or test setups in the base `tsconfig` files.

Happy to be wrong about that, but I don't want tests in our prod outputs and I'm not sure they won't be if they are included here.

File: libs/node/jest.config.js [hunk 9]
```diff
@@ -2,13 +2,11 @@ const { pathsToModuleNameMapper } = require("ts-jest");
 
 const { compilerOptions } = require("../shared/tsconfig.libs");
 
+const sharedConfig = require("../shared/jest.config.base");
+
 module.exports = {
+  ...sharedConfig,
   preset: "ts-jest",
-  testMatch: ["**/+(*.)+(spec).+(ts)"],
-  setupFilesAfterEnv: ["<rootDir>/spec/test.setup.ts"],
```
Comment (MGibson1) [inline libs/node/jest.config.js:8]: I'd like to keep this file around, even though it's empty, as a placeholder for how we initialize cross-testing concerns

**Review by MGibson1 — CHANGES_REQUESTED** [review]
I'm concerned about including spec and test setup files in `tsconfig`. I _think_ webpack should always treeshake these out in production builds, but I'm not 100% confident in my knowledge of that witchcraft.

regardless, we have `tsconfig.spec.json` to handle these -- does that solution not work for some reason? Jest should be honoring those...

File: libs/common/jest.config.js (hunk 6 above)
Comment (eliykat) [reply libs/common/jest.config.js:9]: Good catch!

**Comment by eliykat** [comment]
I've had some wins and losses here 😅 

Win: ts-jest didn't know to use `tsconfig.spec.json`. Once I added that to the base jest config, I could move `files: ["./test.setup.ts"]` into `tsconfig.spec.json` and it still works. This is required to make sure that the setup gets compiled when running tests, but at least it gets it out of the main `tsconfig.json`.

Lose: as discussed, we need to include the `.spec` code in the TS compilation to get path mapping, but VsCode doesn't recongise `tsconfig.json` by any other name. e.g. see  [here](https://stackoverflow.com/questions/57044697/how-to-use-different-tsconfig-file-for-test-files-in-vscode), [here](https://github.com/microsoft/TypeScript/issues/8435). It looks like you can have a separate `tsconfig` in the root of the `spec` directory and that should work, but we're moving away from spec dirs. So we do need to add it to `includes`. We already do this in `libs/common` without a problem, and I inspected the webpack bundle for `web` using `webpack-bundle-analyzer` and this didn't cause any spec files to be added. Which makes sense, because there are no entry points to that code.

After these changes, `web` is the only app to have `.spec` files added to `tsconfig`, so I haven't checked the others. This seems safe for now though.

File: libs/common/tsconfig.json (hunk 7 above)
Comment (eliykat) [reply libs/common/tsconfig.json]: Removed

File: libs/components/tsconfig.json (hunk 8 above)
Comment (eliykat) [reply libs/components/tsconfig.json]: Removed

File: apps/browser/tsconfig.json (hunk 3 above)
Comment (eliykat) [reply apps/browser/tsconfig.json]: Removed

**Review by MGibson1 — APPROVED** [review]
:shipit:

**Pull request merged**